numpy/core/src/multiarray/common.c

```diff
@@ -311,7 +310,6 @@ PyArray_DTypeFromObjectHelper(PyObject *obj, int maxdims,
             dtype = PyArray_DescrNewFromType(NPY_VOID);
             dtype->elsize = buffer_view.itemsize;
             PyBuffer_Release(&buffer_view);
-            _dealloc_cached_buffer_info(obj);
```

Comment:
This really only makes a difference for arrays and scalars... Both should be checked before here.